pkg/server/image_pull.go

```diff
@@ -26,195 +26,223 @@ import (
+//
+// 4) Is the content important if we cached necessary information in-memory
+// after we pull the image? How to manage the disk usage of contents? If some
+// contents are missing but snapshots are ready, is the image still "READY"?
+
 // PullImage pulls an image with authentication config.
 // TODO(mikebrow): add authentication
 // TODO(mikebrow): harden api (including figuring out at what layer we should be blocking on duplicate requests.)
 func (c *criContainerdService) PullImage(ctx context.Context, r *runtime.PullImageRequest) (retRes *runtime.PullImageResponse, retErr error) {
 	glog.V(2).Infof("PullImage %q with auth config %+v", r.GetImage().GetImage(), r.GetAuth())
 	defer func() {
 		if retErr == nil {
-			glog.V(2).Infof("PullImage returns image reference %q", retRes.GetImageRef())
+			glog.V(2).Infof("PullImage %q returns image reference %q",
+				r.GetImage().GetImage(), retRes.GetImageRef())
 		}
 	}()
 
-	var (
-		size int64
-		desc imagespec.Descriptor
-	)
-
-	image, err := normalizeImageRef(r.GetImage().GetImage())
+	namedRef, err := normalizeImageRef(r.GetImage().GetImage())
 	if err != nil {
 		return nil, fmt.Errorf("failed to parse image reference %q: %v", r.GetImage().GetImage(), err)
 	}
+	// TODO(random-liu): [P0] Avoid concurrent pulling/removing on the same image reference.
+	image := namedRef.String()
 	if r.GetImage().GetImage() != image {
-		glog.V(4).Info("PullImage using normalized image ref: %q", image)
-	}
-
-	if desc, size, err = c.pullImage(ctx, image); err != nil {
-		return nil, fmt.Errorf("failed to pull image %q: %v", image, err)
-	}
-	digest := desc.Digest.String() // TODO(mikebrow): add truncIndex for image id
-
-	// TODO(mikebrow): pass a metadata struct to pullimage and fill in the tags/digests
-	// store the image metadata
-	// TODO(mikebrow): consider what to do if pullimage was called and metadata already exists (error? udpate?)
-	meta := &metadata.ImageMetadata{
-		ID:          digest,
-		RepoTags:    []string{image},
-		RepoDigests: []string{digest},
-		Size:        uint64(size), // TODO(mikebrow):  compressed or uncompressed size? using compressed
-	}
-	if err = c.imageMetadataStore.Create(*meta); err != nil {
-		return &runtime.PullImageResponse{ImageRef: digest},
-			fmt.Errorf("pulled image `%q` but failed to store metadata for digest: %s err: %v", image, digest, err)
+		glog.V(4).Infof("PullImage using normalized image ref: %q", image)
 	}
 
-	// Return the image digest
-	return &runtime.PullImageResponse{ImageRef: digest}, err
-}
-
-// normalizeImageRef normalizes the image reference following the docker convention. This is added
-// mainly for backward compatibility.
-func normalizeImageRef(ref string) (string, error) {
-	named, err := reference.ParseNormalizedNamed(ref)
+	// TODO(random-liu): [P1] Schema 1 image is not supported in containerd now, we need to support
+	// it for backward compatiblity.
+	cfgDigest, digest, err := c.pullImage(ctx, image)
```

Comment:
Let's have consistent naming for these two. Maybe `cfgDigest` and `manifestDigest`?
`digest` itself means very little...